src/Lean/Meta/Tactic/AC/Sharing.lean

```diff
@@ -0,0 +1,303 @@
+/-
+Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Alex Keizer
+-/
+prelude
+import Lean.Meta.Tactic.AC.Main
+import Init.Grind.Lemmas
+
+open Lean Meta
+
+/-! ### Types -/
+
+abbrev VarIndex := Nat
+
+structure VarState where
+  varIndices : Std.HashMap Expr VarIndex := {}
+  varExprs : Array Expr := #[]
+
+/-!
+We don't verify the state manipulations, but if we would, these are the invariants:
+```
+structure LegalVarState extends VarState where
+  h_size  : varExprs.size = varIndices.size := by omega
+  h_elems : ∀ h_lt : i < varExprs.size, varIndices[varExprs[i]]? = some i
+```
+-/
+
+abbrev CoefficientsMap := Std.HashMap VarIndex Nat
+
+/-! ### VarState monadic boilerplate  -/
+
+abbrev VarStateM  := StateT VarState MetaM
+
+def VarStateM.run' (x : VarStateM α) (s : VarState := {}) : MetaM α :=
+  StateT.run' x s
+
+/-! ### Implementation -/
+
+/-- Return a list with all variable indices that have a mapping.
+
+Note that this is always a complete sequence `0, 1, ..., (n-1)`, without skipping
+numbers. -/
+def getAllVarIndices : VarStateM (List VarIndex) := do
+  pure <| List.range (← get).varIndices.size
```

Comment:
I don't think we should materialize this list, since we only use it for iteration!

Reply:
I refactored the definition to return a `Range`